napari/components/viewer_model.py

```diff
@@ -558,12 +552,72 @@ def add_layer(self, layer: layers.Layer) -> layers.Layer:
         layer.name = self.layers._coerce_name(layer.name, layer)
         layer.events.name.connect(self.layers._update_name)
         layer.selected = True
-        self.layers.append(layer)
+        # For the labels layer we need to reset the undo/ redo
+        # history whenever the displayed slice changes. Once
+        # we have full undo/ redo functionality, this can be
+        # dropped.
+        if hasattr(layer, '_reset_history'):
+            self.dims.events.ndisplay.connect(layer._reset_history)
+            self.dims.events.order.connect(layer._reset_history)
+            self.dims.events.current_step.connect(layer._reset_history)
+
         self.layers.unselect_all(ignore=layer)
         self._update_layers(layers=[layer])
 
         if len(self.layers) == 1:
             self.reset_view()
+
+        self._on_layers_change(None)
+        self._on_grid_change(None)
+
+    def _on_remove_layer(self, event):
+        """Disconnect old layer events.
+
+        Parameters
+        ----------
+        layer : :class:`napari.layers.Layer`
+            Layer to add.
+
+        Returns
+        -------
+        layer : :class:`napari.layers.Layer` or list
+            The layer that was added (same as input).
+        """
+        layer = event.value
+        layer.events.disconnect()
+        for em in layer.events.emitters.values():
+            em.disconnect()
+        # Disconnect dims events. Once layer._dims removed this
+        # can be removed too.
+        layer._dims.events.disconnect()
+        for em in layer._dims.events.emitters.values():
+            em.disconnect()
+
+        self._on_layers_change(None)
+        self._on_grid_change(None)
+
+    def add_layer(self, layer: layers.Layer) -> layers.Layer:
+        """Add a layer to the viewer.
+
+        Parameters
+        ----------
+        layer : :class:`napari.layers.Layer`
+            Layer to add.
+
+        Returns
+        -------
+        layer : :class:`napari.layers.Layer` or list
+            The layer that was added (same as input).
+        """
+        warnings.warn(
+            (
+                "The viewer.add_layer method is deprecated and will be removed after version 0.4.4."
+                " Instead you should use viewer.layers.append."
+            ),
+            category=DeprecationWarning,
+            stacklevel=2,
+        )
```

Comment:
I do not like this change until _layer group_ feature is implemented. There may be big changes in API of layer keeping. 

Also in semantic meaning, `add_layer` is better than `append`

Reply:
Hmm, that's a good point, we might need to change this for layer groups again, and so we can wait on the deprecation until we know more about how the API might settle, we have some in progress work in #1502

Reply:
I agree with @Czaki here, let's keep `add_layer` for a while longer.

Reply:
ok, keeping it